docs/adr/0003-threadsafe-interfaces.md

```diff
@@ -0,0 +1,196 @@
+# Let consumers opt out of HandleMap locking if their interface is threadsafe
+
+* Status: proposed
+* Deciders: rfkelly, jhugman, mhammond, dmosedale
+* Date:  2021-01-13
+
+## Context and Problem Statement
+
+Uniffi currently uses a very coarse locking strategy for managing concurrent access to object instances,
+which has caused us to accidentally ship code in a product that [blocked the main thread on network I/O](https://jira.mozilla.com/browse/SDK-157).
+We need to enable finer-grained concurrency control in order to provide the desired API for a key consumer.
+
+Currently, every interface has a corresponding [ffi_support::ConcurrentHandleMap](https://docs.rs/ffi-support/0.4.0/ffi_support/handle_map/struct.ConcurrentHandleMap.html) that is responsible for owning all instances of
+that interface and for handing out references to them in a mutability-safe and threadsafe manner. This
+ensures that the generated code is safe in the face of concurrent operations, but has a substantial
+runtime cost: only one method call can be executed on an instance at any time. Any attempt to call an
+object method while a concurrent method is already executing, will block until the previous call has completed.
+
+The desired API for Project Nimbus includes methods that will be called synchronously from the main thread,
+and hence must not block on network or disk I/O. Such an API cannot be built with uniffi as currently
+implemented.
+
+## Decision Drivers <!-- optional -->
+
+* Enabling consumers to control the potential blocking behaviour of their generated APIs.
+* Ensure safety of the generated code.
+* Ship a solution in a timely manner to unblock Project Nimbus.
+
+## Considered Options
+
+* Option A: Do nothing, require consumers to assume that all method calls might block.
+* Option B: Let consumers mark interface definitions as threadsafe to opt in to a non-locking handlemap.
```

Comment:
I had a mild reaction to "non-locking handlemap" as it might imply no locks are necessary at all, but I expect locks will be necessary to protect the handlemap itself? Not sure it's worth trying to squeeze this in though.